### DISCUSSION TITLE
Version numbers / Frequent release feedback

### DISCUSSION THREAD

**Discussion by zaafar** [ocornut/imgui#6507] [New Users Build/Link/Run issues ONLY!]
While I really love the idea of frequent releases and want imgui to move in that direction. it have, unfortunately, created a minor issue for people like me who are releasing imgui.net (this might also be true for other language bindings).

==background==
At Imgui.net side we try to remain in-sync with the imgui version so, once 1.89.6 is released we will try to release imgui.net 1.89.6 as well. There were times when we got out-of-sync but we always try to bring it back to being sync. Due to this (unfortunately) we have to pay big price (e.g. new features/bug fixes/changes on imgui.net side might appear under a patch version)

== why old way worked for us ==
previously imgui didn't use patch versions a lot (e.g. after 1.86, there was 1.87, then there was 1.88 and so now) so we could use patch version for our own features/changes and once imgui releases a new minor version we could sync up again

== why new way is not working for us ==
Now imgui uses patch version (e.g. 1.89.1, 1.89.2 ... ) a lot so we can't use them for our features/bug-fixes/stuff :)

== Questions ==
1: is there a reason why we have started doing that? i.e. rely more on patch versions rather than minor versions?

2: in the new version system, is there a chance of using `a, b, c`? (for example  VERSION 1.33b (2015-02-23))

**Comment by zaafar** [comment]
> 2: in the new version system, is there a chance of using a, b, c? (for example VERSION 1.33b (2015-02-23))

(or for example version 1.89.7b)

**Comment by nicolasnoble** [comment]
FWIW, I have the same issue myself for other packages, and I'm using a
letter-based sub-version number for my own changes. If you were to adopt
this, then in your case, say imgui releases 1.89.3, you release imgui.net
1.89.3a, and if you need to do some patches on it later, you'd release
1.89.3b, 1.89.3c, etc.

**Comment by zaafar** [comment]
yeah that's what i am thinking.

**Comment by zaafar** [comment]
(unfortunately nuget (c# package manager) doesn't like 1.89.3a or 1.89.3b.
it requires a `-` in between (e.g. 1.89.3-a, 1.89.3-rc1, 1.89.3-alpha1) and consider any version with `-` as a pre-release version.

**Comment by emoon** [comment]
What I do in similar situations is just to have my versions fully separated from the library version and when I release a new version of the library I just include in the release notes which version of the library that has been updated to. 

Yes, users will have to read the release notes, but imo it's better as I'm allowed to change my lib version in any way I want.

**Comment by nicolasnoble** [comment]
On nuget, I'm using a 4th version number, so for example, 1.89.3.1,
1.89.3.2, 1.89.3.3, etc.

**Comment by zaafar** [comment]
yeah, I am going to use that as well.

since that is working (Pushing ImGui.NET.1.89.5.1.nupkg to 'https://www.nuget.org/api/v2/package'...)

thank you for all the feedback!

Closing this discussion.

FYI: @ocornut

**Comment by ocornut** [comment]
> 1: is there a reason why we have started doing that? i.e. rely more on patch versions rather than minor versions?

To clarify a few things:

Dear ImGui is not following semantic versioning. I don't believe it is appropriate for this project. So me adding that extra `.x` to the version wasn't intended to mean anything specific (e.g. "patch" in the semantic-versioning scheme of thing).

The truth behind this is: I wrongfully perceived the "xx" in "1.xx" as a % rate leading to 2.0.
So to me after 1.99 would come 2.0. And because I wanted 1.90 and 2.00 to be meaningful releases (in the sense of adding some desirable feature, especially for 2.0) and we are approaching those, I put an arbitrary brake in release numbers. For 1.90 I wanted to finish range-select feature and because that took a while and I accumulated other changes I had to make more intermediary releases.

But this reasoning is flawed, after reading this thread I came to the realization it was perfectly fine to have 1.101, 1.102, 1.156 etc and not immediately go to 2.0. By this logic I think I probably make a mistake in my versioning and 1.89.1-1.89.6 could perfectly have 1.90-1.95.

Depending on the status of range-select work I might adsorb that and not care about 1.90 being a particularly cool release.

Either way, I think your own patching on a repackaged release should add an extra `.1`.

**Reply by ocornut** [reply]
> But this reasoning is flawed, after reading this thread I came to the realization it was perfectly fine to have 1.101, 1.102, 1.156 etc and not immediately go to 2.0. By this logic I think I probably make a mistake in my versioning and 1.89.1-1.89.6 could perfectly have 1.90-1.95.

Addenum: one issue with allowing 1.101 etc is the way we define `IMGUI_VERSION_NUM`.

```
#define IMGUI_VERSION       "1.89.6"
#define IMGUI_VERSION_NUM   18960
```

We would need at one point to bump number of digits `18960 -> `108960` which is fine backward-compatibility wise.

so 1.99 would be 19900 or 109900 (depending on when we bump)
so 1.101 would be 110100

**Reply by zaafar** [reply]
> #define IMGUI_VERSION_NUM   18960


Not saying we should do this but if you go the route of 1.101.2, the chances of reaching 1.101.9+ is low.


So:

1.99.2 can be   010992
1.101.2 can be  011012
10.99.2 can be  100992
10.101.2 can be 101012

**Reply by ocornut** [reply]
That would break #if IMGUI_VERSION_NUM >= checks.

**Reply by zaafar** [reply]
I see, good point. So basically let's continue with xx.yy.zz format.


For now I am good with that since I can basically add xx.yy.zz.kk on my side to solve all my concerns.

**Comment by ocornut** [comment]
Please please also don't use "Discussions" for this. It is stated in bold to not post there for such topic.
(Discussions are gamified and flawed in many ways I am trying to centralize things in Issues for the lack of a better alternative).

I am surprised you intently went past that wall to post your meaningful message :)

(Screenshot)
![image](https://github.com/ocornut/imgui/assets/8225057/205f2fc2-67d1-4026-8577-b8ec62a1694a)

**Reply by zaafar** [reply]
That is my bad! sorry. won't happen again.
